exercises/linked-list/src/example.c

```diff
@@ -1,148 +1,111 @@
-#include <stdlib.h>
-#include <stdbool.h>
 #include "linked_list.h"
+#include <assert.h>
+#include <stdlib.h>
 
-struct list_item {
-   struct list_item *next;
-   struct list_item *prev;
+struct list_node {
+   struct list_node *prev, *next;
    ll_data_t data;
 };
+struct list {
+   struct list_node *first, *last;
+};
 
-static struct list_item *new_item(ll_data_t data)
-{
-   struct list_item *item = malloc(sizeof(*item));
-
-   if (item != NULL) {
-      item->next = NULL;
-      item->prev = NULL;
-      item->data = data;
-   }
-   return item;
-}
-
-static struct list_item *get_list_tail(struct list_item **list)
-{
-   struct list_item *tail = NULL;
-   if (list != NULL) {
-      // climb list to tail
-      tail = *list;
-      while (tail->next != NULL) {
-         tail = tail->next;
-      }
-   }
-   return tail;
-}
-
-static void delete_item(struct list_item *item)
+static struct list_node *list_node_create(struct list_node *prev,
+                                          struct list_node *next,
+                                          ll_data_t data)
 {
-   if (item != NULL) {
-      free(item);
+   struct list_node *node = malloc(sizeof(*node));
+   if (node != NULL) {
+      node->prev = prev;
+      node->next = next;
+      node->data = data;
    }
+   return node;
 }
 
-struct list_item **new_list(void)
+struct list *list_create(void)
 {
-   struct list_item **list = malloc(sizeof(struct list_item *));
+   struct list *list = malloc(sizeof(*list));
    if (list != NULL) {
-      *list = NULL;
+      list->first = NULL;
+      list->last = NULL;
    }
    return list;
 }
 
-bool is_list_empty(struct list_item ** list)
+bool list_is_empty(const struct list *list)
 {
-   return (list == NULL || *list == NULL);
+   assert(list);
+   return list->first == NULL;
 }
 
-bool push(struct list_item ** list, ll_data_t item_data)
+struct list_node *list_push(struct list *list, ll_data_t data)
 {
-   bool result = false;
-   if (list != NULL) {
-      struct list_item *item = new_item(item_data);
-      if (item != NULL) {
-         if (*list != NULL) {
-            struct list_item *old_tail = get_list_tail(list);
-            // append new tail
-            old_tail->next = item;
-            item->prev = old_tail;
-         } else {
-            // add new tail as only item in list
-            *list = item;
-         }
-         result = true;
-      }
-   }
-   return result;
+   assert(list);
+   struct list_node *node = list_node_create(list->last, NULL, data);
+   if (!node)
```

Comment:
Minor nitpick: you're a bit inconsistent in how you check for null. I'd either always use `list` and `!list` or `list != NULL` and `list == NULL`.

Reply:
Done.